components/node-labeler/cmd/run.go

```diff
@@ -249,6 +262,150 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
 	return reconcile.Result{}, nil
 }
 
+type WorkspaceCountController struct {
+	client.Client
+}
+
+func NewWorkspaceCountController(client client.Client) (*WorkspaceCountController, error) {
+	return &WorkspaceCountController{
+		Client: client,
+	}, nil
+}
+
+func (wc *WorkspaceCountController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("workspace-count").
+		For(&workspacev1.Workspace{}).
+		WithEventFilter(workspaceFilter()).
+		Complete(wc)
+}
+
+func workspaceFilter() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			ws := e.Object.(*workspacev1.Workspace)
+			if ws.Status.Runtime == nil {
+				log.WithField("workspace", ws.Name).Info("workspace not ready yet")
+				return false
+			}
+
+			return ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			wsOld := e.ObjectOld.(*workspacev1.Workspace)
+			ws := e.ObjectNew.(*workspacev1.Workspace)
+			if wsOld.Status.Runtime == nil && ws.Status.Runtime != nil {
```

Comment:
Can node name ever be empty for an update? Might be good to check for != "", to be safe.

For example, the workspace's pod could be Pending (not assigned to a node), maybe over the course of a few updates (maybe scale-up is taking a while).

Reply:
You're right, we should definitely check for whether the `NodeName` is actually there

We conditionally assign it, so it's safer if we do the check as you're proposing. 
https://github.com/gitpod-io/gitpod/blob/bd36bac035c22a7b519f3e895c040aaa17916fa0/components/ws-manager-mk2/controllers/status.go#L106-L108